src/main/scala/com/Alteryx/sparkGLM/LM.scala

```diff
@@ -0,0 +1,179 @@
+package com.Alteryx.sparkGLM
+
+import breeze.linalg._
+import breeze.numerics._
+import edu.berkeley.cs.amplab.mlmatrix
+import edu.berkeley.cs.amplab.mlmatrix._
+import edu.berkeley.cs.amplab.mlmatrix.{NormalEquations, RowPartitionedMatrix}
+import edu.berkeley.cs.amplab.mlmatrix.util.Utils
+import edu.berkeley.cs.amplab.mlmatrix.NormalEquations._
+
+import org.apache.spark.{SparkContext, SparkException}
+import org.apache.spark.SparkContext._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.util.Utils
+import org.apache.spark.sql.SQLContext
+import org.apache.spark.sql.{DataFrame, Row}
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.functions
+
+case class PreLM(coefs: DenseMatrix[Double],
+  xtxi: DenseMatrix[Double],
+  sse: Double,
+  r2: Double,
+  fStat: Double)
+
+case class LM(xnames: Array[String],
+              yname: String,
+              coefs: DenseMatrix[Double],
+              stdErr: Array[Double],
+              sigma: Double,
+              r2: Double,
+              fStat: Double,
+              nrow: Double)
+
+object LM {
+
+  // Get X'X and X'y when the DataFrames are row partitioned
+  private def rowPartitionedComponents(
+      X: RowPartitionedMatrix,
+      Y: RowPartitionedMatrix): (DenseMatrix[Double], DenseMatrix[Double])  = {
+    val XY = X.rdd.zip(Y.rdd).map(x => (x._1.mat, x._2.mat))
+    val ATA_ATb = XY.map { part =>
+      (part._1.t * part._1, part._1.t * part._2)
+      }
+
+      val treeBranchingFactor = X.rdd.context.getConf.getInt("spark.mlmatrix.treeBranchingFactor", 2).toInt
+      val depth = math.ceil(math.log(ATA_ATb.partitions.size)/math.log(treeBranchingFactor)).toInt
+      val reduced = edu.berkeley.cs.amplab.mlmatrix.util.Utils.treeReduce(ATA_ATb, utils.reduceNormal, depth=depth)
+
+      reduced
+  }
+
+
+// Figure out how to get (pred - yMean)'(pred - yMean) and (y - yMean)'(y - yMean)
+  // Calculate the sum of the squared errors for the case of partitioned data
+  private def rowPartitionedSSE(
+          X: RowPartitionedMatrix,
+          Y: RowPartitionedMatrix,
+          coefs: DenseMatrix[Double],
+          nrows: Double,
+          ncols: Double): (Double, Double, Double) = {
+    val ySums = Y.rdd.map(y => sum(y.mat(::, 0)))
+    val yMean = (ySums.collect.reduce(_+_))/nrows
+    val XY = X.rdd.zip(Y.rdd).map(x => (x._1.mat, x._2.mat))
+    // the mega tuple goes here
+    val errTopBot = XY.map(x => ((x._2 - (x._1 * coefs)), ((x._1 * coefs) :+ (-1.0*yMean)), (x._2 :+ (-1.0*yMean)))).collect
```

Comment:
This may just be premature optimization on my part, but I think we should take another look at this section to see if we can't streamline things. It goes through all the work of creating a `tuple3` only to then break each piece of the tuple back into a separate `val` assignment, which then gets passed through the same map/reduce pattern 3 times and that worries me.